_posts/news/2025-01-07-announcing-scalajs-1.18.0.md

```diff
@@ -0,0 +1,139 @@
+import scala.util.control.Breaks._
+
+class Parent(val i: Int) {
+  def this(p: Parent) = this(p.i)
+  if (i > 40) throw new IllegalArgumentException
+}
+class Child extends Parent({
+  val p = try new Parent(42) catch { case _: IllegalArgumentException => break() }
+  p
+})
+
+object Test extends App {
+  breakable(new Child)
+}
+{% endhighlight %}
+
+If your application depends on a library that contained such a code shape, you may hit the following IR checking error at link time in Scala.js 1.18.0.
+
+{% highlight none %}
+Restricted use of `this` before the super constructor call
+{% endhighlight %}
+
+If that is the case, the library needs to be recompiled with Scala.js 1.18.0 or later to be usable again.
+Make sure to [tell us on the relevant PR](https://github.com/scala-js/scala-js/pull/5019) if that happens.
+
+In order to mitigate the impact of this change, we conducted an audit of (the latest version of) all Scala.js libraries published to Maven Central.
+We did not find any library that violated the new rule.
+You may still run into the issue if you depend on a private library.
+
+### Drop support for Scala 2.12.2 to 2.12.5 and 2.13.0 to 2.13.2
+
+In order to reduce our maintenance burden and ease further development, Scala.js 1.18.0 drops support for Scala 2.12.2 to 2.12.5 and 2.13.0 to 2.13.2.
+Scala 2.12.6+ and 2.13.3+ are still supported.
+
+We still had a number of dedicated code paths, special cases and exceptions for these versions in our codebase.
+Moreover, we occasionally hit compiler bugs in early 2.12.x versions, which imposed constraints on how we wrote our code.
+
+### The Wasm backend now requires full support of `try_table` (notably Node.js 23)
+
+If you are using the WebAssembly backend, you now need at least Node.js 23 to run and test.
+
+In Scala.js 1.17.0, the WebAssembly backend worked around some issues in Wasm engines regarding the `try_table` instruction.
+Our main blocker was the version of V8 shipped with Node.js 22.
+Now that Node.js 23 has been released with the correct behavior, we removed the workaround.
+
+### Changes to the IR and linker APIs
+
+For tooling authors who directly manipulate the IR and linker APIs, there have been some breaking changes in that area.
+This is in line with our versioning policy for the linker APIs.
+
+The changes you are most likely to hit are:
+
+* We renamed `NoType` to `VoidType`, which represents the IR type `void`.
+  `void` is now generally handled more regularly, as any other type of the type system.
+* `VarRef` now directly contains a `LocalName`, rather than a `LocalIdent` itself containing a `LocalName`.
+* `This` is not a separate IR node anymore; instead it is a `VarRef` with the special name `LocalName.This`.
+* `ArrayLength`, `Throw`, and a number of more obscure IR nodes have been merged into `UnaryOp`.
+* `Transformers` does not make the distinction between "statement" and "expression" anymore.
+  This follows from handling `void` more uniformly.
```

Comment:
Just to check: I guess you have deliberately decided against mentioning LinkTimeProperty?

Reply:
Yes ... but now I reconsidered so I added it.